test/extensions/transport_sockets/internal_upstream/internal_upstream_integration_test.cc

```diff
@@ -182,5 +191,399 @@ TEST_F(InternalUpstreamIntegrationTest, BasicFlowWithoutTransportSocket) {
+  setupBootstrapExtension();
+  setupExternalListenerConfig();
+  setupInternalUpstreamConfig();
+  setupInternalListenerCofnig();
+  setupExternalUpstreamConfig();
+}
+
+// Client sends a clear text HTTP CONNECT message to Envoy.
+// After confirms recevied 200 from proxy, client then sends a
+// clear text HTTP GET message, which will hits the DFP filter in the
+// internal listener and trigger DNS resolving.
+
+TEST_P(HttpsInspectionIntegrationTest, HttpClearInHttpTunnelTest) {
+  setupTestConfig();
+  skip_tag_extraction_rule_check_ = true;
+  initialize();
+
+  // Open clear-text connection.
+  conn_->connect();
+
+  // Sends a HTTP CONNECT message.
+  Buffer::OwnedImpl buffer;
+  std::string connect_msg =
+    "CONNECT www.cnn.com:80 HTTP/1.1\r\n"
+    "Host: www.cnn.com:80\r\n"
+    "\r\n\r\n";
+  payload_reader_->set_data_to_wait_for("HTTP/1.1 200 OK");
+  buffer.add(connect_msg);
+  conn_->write(buffer, false);
+  dispatcher_->run(Event::Dispatcher::RunType::Block);
+
+  // Sends a HTTP GET message.
+  auto port_str = std::to_string(fake_upstreams_[0]->localAddress()->ip()->port());
+  std::string get_msg =
+    "GET / HTTP/1.1\r\n"
+    "Host: localhost:"+ port_str  + "\r\n"
+    "\r\n\r\n";
+  buffer.add(get_msg);
+  conn_->write(buffer, false);
+  while (client_write_buffer_->bytesDrained() != connect_msg.length() +
+         get_msg.length()) {
+    dispatcher_->run(Event::Dispatcher::RunType::NonBlock);
+  }
+
+  FakeRawConnectionPtr fake_upstream_connection;
+  ASSERT_TRUE(fake_upstreams_[0]->waitForRawConnection(fake_upstream_connection));
+  ASSERT_TRUE(fake_upstream_connection->waitForData(
+      FakeRawConnection::waitForInexactMatch("GET / HTTP/1.1")));
+  // Sends a response back to client.
+  payload_reader_->set_data_to_wait_for("HTTP/1.1 503 Service Unavailable");
+  ASSERT_TRUE(fake_upstream_connection->write(
+      "HTTP/1.1 503 Service Unavailable\r\n\r\n"));
+  dispatcher_->run(Event::Dispatcher::RunType::Block);
+
+  conn_->close(Network::ConnectionCloseType::FlushWrite);
+}
+
+// Client sends a clear text HTTP CONNECT message to Envoy.
+// After confirms recevied 200 from proxy, client then start TLS handshake.
+// Then Client then sends a HTTP GET message, which will hit DFP filter in the
+// internal listener and trigger DNS resolving.
+//
+// This test is not working yet since the TLS handshake is failing with below error message:
+// remote address:envoy://internal_client_address/,TLS error: 268435703:SSL
+// routines:OPENSSL_internal:WRONG_VERSION_NUMBER
+
+
+TEST_P(HttpsInspectionIntegrationTest, HttpsInHttpTunnelTest) {
+  clear_http_test_ = false;
+  setupTestConfig();
+  skip_tag_extraction_rule_check_ = true;
+  initialize();
+  // Open clear-text connection.
+  conn_->connect();
+  Buffer::OwnedImpl buffer;
+  std::string request_msg =
+    "CONNECT www.cnn.com:80 HTTP/1.1\r\n"
+    "Host: www.cnn.com:80\r\n"
+    "\r\n\r\n";
+  payload_reader_->set_data_to_wait_for("HTTP/1.1 200 OK");
```

Comment:
It is possible that not all response data to CONNECT request was read before trying to switch to TLS. As a result TLS client is trying to interpret leftover bytes from CONNECT response as TLS. It is unlikely but could be source of flakiness down the road.

Strictly speaking you need to wait for `\r\n\r\n` before switching to TLS. What happens if you wait for that sequence. If you want to verify both `HTTP/1.1 200 OK` and wait for `\r\n\r\n` you may need to modify the WaitForPayloadReader to wait until multiple strings match.